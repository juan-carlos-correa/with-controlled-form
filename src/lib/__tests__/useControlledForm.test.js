import React from 'react';
import { mount, shallow } from 'enzyme';
import Form from '../../demo/components/Form';
import { useControlledForm } from '../';

const formState = {
  email: '',
  phoneNumber: '',
};

const formValidations = {
  email: {
    isRequired: true,
    isMinLength: 3,
    isMaxLength: 60,
    isEmail: [true, 'email is not valid!'],
  },
  phoneNumber: {
    isRequired: true,
    isMinLength: 6,
    isMaxLength: 15,
  }
};

const FormWithUseControlledFormHook = ({ myHandleSubmit }) => {
  const {
    values,
    errors,
    handleChange,
    handleSubmit
   } = useControlledForm(formState, formValidations);

  return <Form
    values={values}
    errors={errors}
    handleChange={handleChange}
    handleSubmit={() => handleSubmit(myHandleSubmit)}
  />
}

describe('useControlledForm hook', () => {
  let wrapper = null;
  let handleSubmitMock = null;

  beforeEach(() => {
    handleSubmitMock = jest.fn();
    wrapper = mount(<FormWithUseControlledFormHook myHandleSubmit={handleSubmitMock} />);
  });

  afterEach(() => {
    handleSubmitMock.mockClear();
    wrapper.unmount();
  });

  test('render without crash', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('call submit handler function properly with valid values', () => {
    const email = wrapper.find('#email');
    expect(email).toHaveLength(1);

    email.simulate('change', { target: { value: 'foo@mail.com', name: 'email' } });

    const phoneNumber = wrapper.find('#phoneNumber');
    expect(phoneNumber).toHaveLength(1);

    phoneNumber.simulate('change', { target: { value: '5512345678', name: 'phoneNumber' } });

    const form = wrapper.find('form');
    form.simulate('submit');

    expect(handleSubmitMock).toHaveBeenCalledTimes(1);
  });
})
