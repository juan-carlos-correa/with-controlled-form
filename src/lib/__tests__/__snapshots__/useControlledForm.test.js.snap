// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useControlledForm hook render without crash 1`] = `
<FormWithUseControlledFormHook
  myHandleSubmit={[MockFunction]}
>
  <Form
    errors={
      Object {
        "email": Array [],
        "phoneNumber": Array [],
      }
    }
    handleChange={[Function]}
    handleSubmit={[Function]}
    values={
      Object {
        "email": "",
        "phoneNumber": "",
      }
    }
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="form-element"
      >
        <label
          htmlFor="email"
        >
          Email:
        </label>
        <input
          className="input"
          id="email"
          name="email"
          onChange={[Function]}
          type="text"
          value=""
        />
        <div
          className="error-message"
        />
      </div>
      <div
        className="form-element"
      >
        <label
          htmlFor="phoneNumber"
        >
          Phone number:
        </label>
        <input
          className="input"
          id="phoneNumber"
          name="phoneNumber"
          onChange={[Function]}
          type="tel"
          value=""
        />
        <div
          className="error-message"
        />
      </div>
      <div
        className="text-center"
      >
        <button
          className="button"
          id="send"
          type="submit"
        >
          Send
        </button>
      </div>
    </form>
  </Form>
</FormWithUseControlledFormHook>
`;

exports[`useControlledForm hook useControlledForm hook render without crash 1`] = `
<FormWithUseControlledFormHook
  myHandleSubmit={[MockFunction]}
>
  <Form
    errors={
      Object {
        "email": Array [],
        "phoneNumber": Array [],
      }
    }
    handleChange={[Function]}
    handleSubmit={[Function]}
    values={
      Object {
        "email": "",
        "phoneNumber": "",
      }
    }
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="form-element"
      >
        <label
          htmlFor="email"
        >
          Email:
        </label>
        <input
          className="input"
          id="email"
          name="email"
          onChange={[Function]}
          type="text"
          value=""
        />
        <div
          className="error-message"
        />
      </div>
      <div
        className="form-element"
      >
        <label
          htmlFor="phoneNumber"
        >
          Phone number:
        </label>
        <input
          className="input"
          id="phoneNumber"
          name="phoneNumber"
          onChange={[Function]}
          type="tel"
          value=""
        />
        <div
          className="error-message"
        />
      </div>
      <div
        className="text-center"
      >
        <button
          className="button"
          id="send"
          type="submit"
        >
          Send
        </button>
      </div>
    </form>
  </Form>
</FormWithUseControlledFormHook>
`;

exports[`useControlledForm tests useControlledForm hook render without crash 1`] = `
<FormWithUseControlledFormHook
  myHandleSubmit={[MockFunction]}
>
  <Form
    errors={
      Object {
        "email": Array [],
        "phoneNumber": Array [],
      }
    }
    handleChange={[Function]}
    handleSubmit={[Function]}
    values={
      Object {
        "email": "",
        "phoneNumber": "",
      }
    }
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="form-element"
      >
        <label
          htmlFor="email"
        >
          Email:
        </label>
        <input
          className="input"
          id="email"
          name="email"
          onChange={[Function]}
          type="text"
          value=""
        />
        <div
          className="error-message"
        />
      </div>
      <div
        className="form-element"
      >
        <label
          htmlFor="phoneNumber"
        >
          Phone number:
        </label>
        <input
          className="input"
          id="phoneNumber"
          name="phoneNumber"
          onChange={[Function]}
          type="tel"
          value=""
        />
        <div
          className="error-message"
        />
      </div>
      <div
        className="text-center"
      >
        <button
          className="button"
          id="send"
          type="submit"
        >
          Send
        </button>
      </div>
    </form>
  </Form>
</FormWithUseControlledFormHook>
`;
